### ISSUE TITLE
Add yields in the summary process to allow user interaction to go through

### ISSUE THREAD

**Issue by agarwal-navin** [microsoft/FluidFramework#3969]
Adding yields in the summary process will make it slightly less priority and allow user interactions to go through. With garbage collection being part of the summarize process, it will take more time and we want to make the runtime responsive during this process.

**Comment by vladsud** [comment]
The goal of this task is to add yielding into the process.
```typescript
export const yieldExecution = async () => new Promise((accept) => setTimeout(accept, 0));
```
And then sprinkling 
```typescript
await yieldExecution ();
```
in various places in the code.

For example, likely first (and only) starting point to add that yield would be in FluidDataStoreContext.summarizeInternal(), i.e. give foreground tasks a chance to execute some code in between summarizer processing data stores.

Note that in worst case, we slow down summarization process by 16ms (minimum scheduling quantum in OS) * number of data stores that need to be summarized. This should be rather small number assuming normal (incremental) summaries.
Full (safe) summary could have 1K data stores easily.

Some pre-reading might be useful on how timers work in JS:
https://johnresig.com/blog/how-javascript-timers-work/

@arinwt - can you please glance and critique?

**Comment by arinwt** [comment]
This makes sense to me. Do we already see perf issues for the oldest client when forming a big summary?

For something like this it might be nice to have an object we can reference for time-based yields. Something where we start a timer and invalidate a flag every `n` ms, then yield at next opportunity. Some sample code:

```ts
const minYieldWaitTime = 1000; // do not yield more frequently than once per 1 second

const yieldTracker = {
    needsYield: false,
    finished: false,
};
const shouldYieldHandler = () => {
    if (!yieldTracker.finished) {
        yieldTracker.needsYield = true;
        setTimeout(shouldYieldHandler, minYieldWaitTime);
    }
};
const shouldYieldTimer = setTimeout(shouldYieldHandler, minYieldWaitTime);

try {
    await doLongRunningWork(yieldTracker);
} finally {
    yieldTracker.finished = true; // could also be set somewhere in long running work
}
```

Where the long running work has some checks in it like so:
```ts
async doLongRunningWork(yieldTracker: { needsYield: boolean; finished: boolean; }) {
    for (let i = 0; i < 1000; i++) {
        if (yieldTracker.needsYield) {
            await yieldExecution();
        }
        // do work
    }
}
```

Then we could more safely call the needsYield check while summarizing at each data store or dds, and only await if needed. This would of course be cleaned up, encapsulated, and maybe use our Timer objects.

**Comment by vladsud** [comment]
One thing to consider - I believe yielding in data store layer is likely noop, because I believe we would do Promise.all() somewhere when kicking off summaries for each data store (this code was used to be in ConainerRuntime, I believe it's somewhere in summary node). I.e. if they all will yield at the same time.

That's just a theory based on how I think code works :)
If so, yielding needs to move to some other layer?

**Comment by vladsud** [comment]
@arinwt , @agarwal-navin, can you please help here with more detailed design?
For example, it's not very clear where such yields to be inject (I believe doing it in data store context / runtime will not be sufficient due to Promise.All() kind of activity around all of data stores), i.e. "concurrency".

As for data - yes, we know that Whiteboard summarized a ton of data and it does cause hangs. Some of that is their problems to solve, some of it - we do not provide good tools in other areas (like ability for DDSs to reuse blobs from prior summaries), but some of it - 99th percentile will always be CPU intensive.

That said, we need to figure out how to get better at instrumenting this. I assume we already have time measurements and it tells us half the story (wall clock time, not CPU time). Can we improve it? Can we maintain what we got after injecting yields (i.e. measure number of yields and their total duration such that we can continue to measure wall clock time of actual summarization)?

**Comment by agarwal-navin** [comment]
@vladsud, we can pick this up for July. Arin and I can look into the design for this.

**Comment by arinwt** [comment]
It does use `Promise.all()`, but I don't think that necessarily causes a problem by itself: each promise will synchronously execute, but as long as the first line in each promise is checking `needsYield`, then they will be no-ops.

Doing it at the datastore layer would probably _not_ be sufficient though. It would work for a large number of datastores, but not for example if there are many DDS's or a single complicated DDS. We can inject the `needsYield` in the datastore runtime to solve the first problem (many DDS's), but for single complicated DDS, we would need to propagate the yield tracker down to the DDS itself.

Perhaps a better implementation of the yield tracker would be diffing the timestamp in the `needsYield()` check. We would then need a way to ensure both that all yield checks are respected (we can't synchronously make `needsYield()` false), and that all yield checks are guaranteed a chance to run (we can't let them perpetually be blocked by external code taking longer than minYieldTime). I think the solution to both of these lies in the nuance of `yield()` below:

```ts
const minYieldWaitTime = 1000; // do not yield more frequently than once per 1 second

class YieldTracker {
    private lastYieldTime: number;
    private yieldPromise: Promise<void> | undefined;

    constructor(private readonly minYieldTime: number, private readonly now = performance.now) {
        this.lastYieldTime = this.now();
    }

    public get needsYield() {
        if (this.lastYieldTime === undefined) {
            return false;
        }
        const elapsed = this.now() - this.lastYieldTime;
        return elapsed > this.minYieldTime;
    }

    public async yield() {
        if (this.yieldPromise === undefined) {
            this.yieldPromise = new Promise((accept) => setTimeout(accept, 0)).then(() => {
                this.lastYieldTime = undefined;
                setTimeout(() => {
                    this.lastYieldTime = this.now();
                });
            });
        }
        return this.yieldPromise;
    }
}
```

Above is sample code (not tested), but the idea is to show the importance of balancing both things mentioned above. It is _not_ a complete solution, because it will allow all yieldable code threads to execute at least one step of their code after the first yield, when ideally we'd want to start the timer ASAP, but it shows that we can't start the timer too quickly.

**Comment by vladsud** [comment]
@arinwt - thanks for the comment! Given that summary nodes control summarization process, is there single bottleneck that can have that kind of calls?

Moving it out to August, as I think we need to have more clarity (and likely a bit more time) for Si to execute on it.

**Comment by pleath** [comment]
Does the presence of async summarization affect the need for this item?

**Comment by vladsud** [comment]
Not really. async generates micro-tasks, but they are all part of same JS turn, unless there are some timeouts or blocking calls (like network calls). In case of summarization, it's all CPU intensive code, so it will take all the time, unless there are some yields in between that use timeouts.
